spec/models/book_spec.rb

```diff
@@ -73,6 +73,35 @@
     end
   end
 
+  describe ".leap_year" do
+    it "Determines whether or not the book was written in a leap year" do
+      book = Book.new
+      book.created_at = Date.new(2000, 1, 1)
+
+      expect(book.leap_year?).to eq(true)
+
+      book.created_at = book.created_at.change(year: 1700)
+
+      expect(book.leap_year?).to eq(false)
+
+      book.created_at = book.created_at.change(year: 2016)
+
+      expect(book.leap_year?).to eq(true)
+
+      book.created_at = book.created_at.change(year: 1900)
+
+      expect(book.leap_year?).to eq(false)
+
+      book.created_at = book.created_at.change(year: 2100)
+
+      expect(book.leap_year?).to eq(false)
+
+      book.created_at = book.created_at.change(year: 2400)
+
+      expect(book.leap_year?).to eq(true)
+    end
```

Comment:
We have some tests that prove some years return true and others return false. Let's take this another step further, I want these tests to teach the reader how leap years work.

Looking at these tests, without jumping directly into the code, I don't know which cases each of these years covers. You have a few specific cases you want to cover, based on your boolean logic in the method.

Let's look at what we have in the leap_year method.

```ruby
(year % 4 == 0 && year % 100 != 0) || (year % 400 == 0)
```

We can see there are two conditions we care about for something to be a leap year:

1. `(year % 4 == 0 && year % 100 != 0)`, and
2. `(year % 400 == 0)`

We want to have the fewest number of years that can prove these rules work.

The rules you have defined in English are:

1.  A year is a leap year if it is divisible by 4 AND not divisible by 100
2. A year is a leap year if it is divisible by 400

So we need to prove those 2 conditions are leap years, and we need to prove that outside of that years are not leap years.

Basically we'll write something like a [truth table](https://en.wikipedia.org/wiki/Truth_table#:~:text=A%20truth%20table%20is%20a,taken%20by%20their%20logical%20variables.).

Let's go, here are the cases we need, and why we need them. I usually write all the positive cases first, followed by the negative cases.

1. Divisible by 4 and not divisible by 100 (EG: 2016) IS a leap year
2. Divisible by 4 and divisible by 100 and divisible by 400 (EG: 2000) IS a leap year
3. Divisible by 4 and divisible by 100 but not divisible by 400 (EG: 1900) IS NOT a leap year
4. Not divisible by 4 (EG: 2015) IS NOT a leap year

I think outside of that, all the numbers we can choose just validate one or another of the rules we've already covered above.

Reply:
Looking at that more from the truth table angle, we have 3 booleans:

A: divisible by 4
B: not divisible by 100
C: divisible by 400

We need to test:

- ABC is true
- A!BC is true
- AB!C is false
- !(ABC) is false

this is a bit of a lopsided truth table because B and C are related - B and C can both be true, but if C is true B must also be true.